File: src/main/java/gg/mineral/server/network/packet/play/clientbound/MultiBlockChangePacket.java
package gg.mineral.server.network.packet.play.clientbound;

import java.util.List;

import gg.mineral.server.network.packet.Packet;
import io.netty.buffer.ByteBuf;
import lombok.val;

public record MultiBlockChangePacket(int chunkX, int chunkZ, List<BlockChangePacket> records)
        implements Packet.OUTGOING {
    @Override
    public void serialize(ByteBuf os) {
        os.writeInt(chunkX);
        os.writeInt(chunkZ);
        os.writeShort(records.size());
        os.writeInt(records.size() * 4);

        for (val record : records) {
            // XZYYTTTM
            int value = (record.blockMetadata() & 0xF) |
                    ((record.blockId() & 0xFFF) << 4) |
                    ((record.y() & 0xFF) << 16) |
                    ((record.z() & 0xF) << 24) |
                    ((record.x() & 0xF) << 28);
            os.writeInt(value);
        }
    }

    @Override
    public byte getId() {
        return 0x22;
    }
}

File: server/src/main/kotlin/gg/mineral/server/network/packet/play/clientbound/MultiBlockChangePacket.kt
package gg.mineral.server.network.packet.play.clientbound

import gg.mineral.api.network.packet.Packet
import io.netty.buffer.ByteBuf

@JvmRecord
data class MultiBlockChangePacket(val chunkX: Int, val chunkZ: Int, val records: List<BlockChangePacket>) :
    Packet.OUTGOING {
    override fun serialize(os: ByteBuf) {
        os.writeInt(chunkX, chunkZ)
        os.writeShort(records.size)
        os.writeInt(records.size * 4)

        for ((x, y, z, blockId, blockMetadata) in records) {
            // XZYYTTTM
            val value = (blockMetadata.toInt() and 0xF) or
                    ((blockId and 0xFFF) shl 4) or
                    ((y.toInt() and 0xFF) shl 16) or
                    ((z and 0xF) shl 24) or
                    ((x and 0xF) shl 28)
            os.writeInt(value)
        }
    }

    override val id: Byte
        get() = 0x22
}
